### ISSUE TITLE
Errors trying the delta-sharing access with pandas and spark

### ISSUE THREAD

**Issue by pknowles-9** [delta-io/delta-sharing#74]
Hello team,

I'm trying to use your delta-sharing library with python for both PANDAS and SPARK

**SPARK METHOD**

As you can see from the below screenshot I can access the columns or schema using the load_as_spark() method

![20211101_154435](https://user-images.githubusercontent.com/93537448/139700299-6f93395e-5c74-4030-834f-82fcd84c4d76.jpg)

However, when I try .show() I see the following Java certificate errors

javax.net.ssl.SSLHandshakeException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target 

Caused by: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target 

I've tried to use different Java versions and also added the certificate of the delta-sharing server hosted on Kubernetes, but nothing seems to be working. If I try this from a linux machine or kubernetes I seem to have the same problem

**PANDAS METHOD**

When I use the load_as_pandas() method I get a FileNotFoundError. The strange thing is that when I click on that s3 url link that you can see from the screenshot, I download the parquet file (therefore the link does work). Is the PyArrow library trying to look for a local file or what do you think the issue may be ?

![20211101_161049](https://user-images.githubusercontent.com/93537448/139703789-9888f6be-5b36-4730-9e0e-72d6c8eeb14f.jpg)

Any ideas about the above 2 errors using spark and pandas ?

Thank you very much,

Peter

**Comment by zsxwing** [comment]
How do you config certificate for the server? Could you add the following configs?
```
ssl:
# The file of the PEM-format certificate
- certificateFile: "..."
# The file of the certificate’s private key
- certificateKeyFile: "..."
# The file storing the password to access the above certificate’s private key if it's protected
- certificatePasswordFile: "..."
```

I realized this was not documented. Feel free to open a PR to update README to add it 😄

**Comment by pknowles-9** [comment]
Thank you zsxwing,

Do you mean that I should add this ssl content to the 'delta-sharing-server.yaml' file  ?

**Unpack the pre-built package and copy the server config template file conf/delta-sharing-server.yaml.template to create your own server yaml file, such as conf/delta-sharing-server.yaml.
Make changes to your yaml file. You may also need to update some server configs for special requirements.**

And what about the pandas issue (FileNotFoundError) ? Have you seen that before ?

Thank you very much,

Peter

**Comment by pknowles-9** [comment]
Run the following shell command:

bin/delta-sharing-server -- --config **<the-server-config-yaml-file>** 
<the-server-config-yaml-file> should be the path of the yaml file you created in the previous step. You can find options to config JVM in sbt-native-packager.

**Comment by pknowles-9** [comment]
Run the following shell command:

bin/delta-sharing-server -- --config **the-server-config-yaml-file**
<the-server-config-yaml-file> should be the path of the yaml file you created in the previous step. You can find options to config JVM in sbt-native-packager.

**Comment by pknowles-9** [comment]
I also have in conf my core-site.xml file with "fs.s3a.access.key" and "fs.s3a.secret.key". I must have access to S3 with my credentials to be able to view columns/schema (for spark) and urls (for pandas)

**Comment by zsxwing** [comment]
> Do you mean that I should add this ssl content to the 'delta-sharing-server.yaml' file ?

Yep. `ssl` configs should be added to `delta-sharing-server.yaml`.

> I also have in conf my core-site.xml file with "fs.s3a.access.key" and "fs.s3a.secret.key".

Yep. Correct. This is supported by Delta Sharing 0.2.0.

**Comment by pknowles-9** [comment]
Thank you zsxwing,

I will try this ssl change

What about the pandas issue (FileNotFoundError) ? Have you seen that before ? Could it also be related to the ssl issue ?

Thank you very much,

Peter

**Comment by zsxwing** [comment]
Ah, I missed it. How did you create this table? Were you using Databricks or OSS Delta Lake?

**Comment by pknowles-9** [comment]
I'm not sure about the difference between Databricks and OSS Delta Lake. To create the pandas dataframe I used the below code. The screenshot with the output is up at the top. The S3 url works and I can click on it and open the parquet file. However, from the error it looks like pyarrow doesn't seem to be able to find this file. I wonder if it could be related to the pyarrow version

##########################################################################

import delta_sharing

# Point to the profile file. It can be a file on the local file system or a file on a remote storage.
profile_file = "<profile-file-path>"

# Create a SharingClient.
client = delta_sharing.SharingClient(profile_file)

# List all shared tables.
client.list_all_tables()

# Create a url to access a shared table.
# A table path is the profile file path following with `#` and the fully qualified name of a table (`<share-name>.<schema-name>.<table-name>`).
table_url = profile_file + "#<share-name>.<schema-name>.<table-name>"

# Load a table as a Pandas DataFrame. This can be used to process tables that can fit in the memory.
delta_sharing.load_as_pandas(table_url)

**Comment by pknowles-9** [comment]
I've just tried the latest version of pyarrow but that doesn't seem to help

**Comment by zsxwing** [comment]
Sorry. I meant the Delta table in your s3 bucket. By the way, it doesn't look like a standard s3 bucket. I don't see the endpoint in https://docs.aws.amazon.com/general/latest/gr/s3.html

**Comment by pknowles-9** [comment]
I'm not sure about that. I think neither of the two (Databricks or OSS Delta Lake). I'll have to ask someone in my team and let you know.

For the delta-sharing library to work properly with pandas, do we have to use a standard s3 bucket ?

If there is something wrong with the s3 bucket, the strange thing is that:

- WITH PANDAS -> the S3 url I get back does work and I can download the parquet file. However, I get this FileNotFoundError
- WITH SPARK -> the S3 works and I can see the schema/columns for example (but not the .show() data due to that possible ssl error) 

Thank you very much,

Peter

**Comment by pknowles-9** [comment]
Hi zsxwing,

I have another question

Are the ssl details only needed when using Python ?

A colleague has managed to get data using Java without providing ssl details

**Comment by zsxwing** [comment]
> do we have to use a standard s3 bucket ?

`FileNotFoundError` is usually because the pre-signed url generated by the user is not correct. If you are using non standard S3 bucket here, it may have a different behavior. I would like to understand the details about the S3 bucket so that I may be able to try to reproduce it by myself.

> Are the ssl details only needed when using Python ?

We don't require to use SSL. You can remove the ssl config from `delta-sharing-server.yaml` and use `http` urls. But without SSL, the communication may be exposed to other people over the network. I'd recommend to use it if possible.

**Comment by pknowles-9** [comment]
Thank you zsxwing,

I've asked my team and they've said it's an s3a bucket

When I try load_as_pandas() from my Linux dev machine (rather than Kubernetes) I get a different error

_________________________________

DeltaSharingReader._to_pandas(file, converters) for file in response.add_files]

dataset(source=add_file.url, format="parquet", filesystem=filesystem)

return

**Comment by pknowles-9** [comment]
Sorry closed by mistake

return _filesystem_dataset(source, **kwargs)

fs, paths_or_selector = ensure_single_source(source, filesystem)

filesystem, _ = ensure_filesystem(filesystem)

'Filesystem argument must be a FileSystem instance or a valid system URI

**Comment by pknowles-9** [comment]
From the Spark system I get the same error regardless of of I try from Kubernetes or my Linux machine:

_____________________

javax.net.ssl.SSLHandshakeException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target

Caused by: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target

__________________________

I think I'm using http rather than https, as you can see from the below image (http://127.0.0.1:8080)

![16363655111203581737064052575578](https://user-images.githubusercontent.com/93537448/140721862-66b82c49-3d23-4d89-b697-b1ab63c4d537.jpg)

This is my delta-sharing-server.yaml

![16363656773162989177358923639135](https://user-images.githubusercontent.com/93537448/140722249-e3894e1d-4af6-4bd5-baec-d0f562ac939a.jpg)

![16363657193827139373644182029200](https://user-images.githubusercontent.com/93537448/140722352-77f68eb1-5a85-4d83-b0f3-f361c7945df8.jpg)

I don't mention ssl as you can see above

This is my profile file

![16363659958047946189603383700806](https://user-images.githubusercontent.com/93537448/140723040-8a6e9727-7407-42b6-89f7-c305f5b0da81.jpg)

They've also told me that the delta tables are not created via Databricks or OSS DeltaLake

They are creating the tables via an API

datasetFrame.write.format("delta")...

**Comment by pknowles-9** [comment]
From Kubernetes the load_as_pandas error is:

![16363663379457053826759052860406](https://user-images.githubusercontent.com/93537448/140723851-12a9410e-9486-47ae-91c3-c5b9fa905a93.jpg)

![16363663782963250526735478065127](https://user-images.githubusercontent.com/93537448/140723925-90bc855c-c097-4764-9293-ec4cd8208106.jpg)

fs, paths_or_selector = _ensure_single_source(source, filesystem)

raise FileNotFoundError (path)

**Comment by pknowles-9** [comment]
I've now got both to work from my Linux machine

Spark was being blocked somehow by the Java permissions at the company

![16364535518745683022200353311135](https://user-images.githubusercontent.com/93537448/140907408-8568c881-bbce-40d4-b28a-cc5ae0b9d2e2.jpg)


Pandas I got it to work updating pyarrow

![16364535130801364049080422220227](https://user-images.githubusercontent.com/93537448/140907315-7ac2e975-9ada-47f7-be4a-b0c4624474cb.jpg)

I'll now need to figure out how to get it to work from Kubernetes. I'll close this issue. 

Thanks!
